### PR TITLE
feat: Implement the task chunking RFC 0001

### PR DESCRIPTION
**Pull request by mwiebe**
### What was the problem/requirement? (What/Why)

https://github.com/OpenJobDescription/openjd-specifications/blob/mainline/rfcs/0001-task-chunking.md

### What was the solution? (How)

Implement the task chunking additions as the TASK_CHUNKING extensions, using the context-sensitive parsing introduced by the Pydantic V2 conversion and extensions RFC implementation.

### What is the impact of this change?

Job templates that use the task chunking feature can be parsed successfully.

### How was this change tested?

Added tests for the new functionality.

- Have you run the unit tests?

Yes

### Was this change documented?

- Are relevant docstrings in the code base updated?

Yes

### Is this a breaking change?

Not a breaking change, it adds support for the TASK_CHUNKING extension.

### Does this change impact security?

This does not impact the threat model, and the new extensions field and parameter follow the same level of input validation relevant for existing threats.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/openjd/model/v2023_09/_model.py [hunk 1]
```diff
@@ -797,11 +851,134 @@ class PathTaskParameterDefinition(OpenJDModel_v2023_09):
...
+                            ctx={
+                                "error": ValueError("Value must be an integer or integer string.")
+                            },
+                            input=item,
+                        )
+                    )
+            if errors:
+                raise ValidationError.from_exception_data(cls.__name__, line_errors=errors)
+        elif isinstance(value, RangeString):
+            # TODO: nothing to do - it's guaranteed to be a format string at this point
```
Comment (leongdl) [inline src/openjd/model/v2023_09/_model.py]: If this isn't a todo, then its just a comment ?

Comment (leongdl) [reply src/openjd/model/v2023_09/_model.py]: IE, we can remove TODO:

File: test/openjd/model/v2023_09/test_create.py [hunk 2]
```diff
@@ -394,3 +396,119 @@ def test(self) -> None:
...
+        #  2) Every entity and every field that exists is defined at least once.
+        #  3) Testing of _internal.create_job covers corner cases & exceptions; we don't worry
+        #     about those here.
+
+        # GIVEN
+        extra_kwargs = {"$schema": "blah "}  # special snowflake due to field naming
+        template = JobTemplate(
+            **extra_kwargs,
+            specificationVersion="jobtemplate-2023-09",
+            extensions=["TASK_CHUNKING"],
```
Comment (leongdl) [inline test/openjd/model/v2023_09/test_create.py:416]: Can we add a templated test to test with an empty array ?

File: src/openjd/model/v2023_09/_model.py (hunk 1 above)
Comment (mwiebe) [reply src/openjd/model/v2023_09/_model.py]: Good point, I'll fix that and where I got it from.

File: test/openjd/model/v2023_09/test_create.py (hunk 2 above)
Comment (mwiebe) [reply test/openjd/model/v2023_09/test_create.py:416]: That case is already tested here: https://github.com/OpenJobDescription/openjd-model-for-python/blob/mainline/test/openjd/model/test_parse.py#L254

This file test_create.py contains larger end-to-end tests for converting an already-validated JobTemplate object into a Job. What error cases are you thinking about?

Comment (leongdl) [reply test/openjd/model/v2023_09/test_create.py:416]: Just thinking to cover the different range of inputs, null (before), [] (empty new), populated (this test case).

Cool, if its covered then all good!

**Review by leongdl — APPROVED** [review]

**Review by epmog — APPROVED** [review]

File: src/openjd/model/v2023_09/_model.py [hunk 3]
```diff
@@ -553,6 +561,52 @@ def _validate_range_expression(cls, value: Any) -> Any:
...
+    _job_creation_metadata = JobCreationMetadata(
+        resolve_fields={"defaultTaskCount", "targetRuntimeSeconds"},
+    )
+
+    @field_validator("defaultTaskCount")
+    @classmethod
+    def _validate_default_task_count(cls, value: Any) -> Any:
+        if isinstance(value, FormatString):
+            # If the string value has no expressions, can validate the value now.
+            # Otherwise will validate when
```
Comment (jusiskin) [inline src/openjd/model/v2023_09/_model.py:585]: nit: this comment is unfinished

File: test/openjd/model/v2023_09/test_chunk_int_task_parameter_type.py [hunk 4]
```diff
@@ -0,0 +1,658 @@
...
+                "type": "CHUNK[INT]",
+                "range": "1-100",
+                "chunks": {
+                    "defaultTaskCount": 10,
+                    "targetRuntimeSeconds": "{{Param.TargetChunkRuntime}}",
+                    "rangeConstraint": "NONCONTIGUOUS",
+                },
+            },
+            id="targetRuntimeSeconds is str expression",
+        ),
```
Comment (jusiskin) [inline test/openjd/model/v2023_09/test_chunk_int_task_parameter_type.py:159]: nit: Should we also have a case where `defaultTaskCount` is a format string? I see it is already covered in `test/openjd/model/v2023_09/test_create.py`, but maybe for completeness it would be good to have here as well.

File: test/openjd/model/v2023_09/test_chunk_int_task_parameter_type.py [hunk 5]
```diff
@@ -0,0 +1,658 @@
...
+                "range": [1.1],
+                "chunks": {
+                    "defaultTaskCount": 10,
+                    "targetRuntimeSeconds": 1000,
+                    "rangeConstraint": "NONCONTIGUOUS",
+                },
+            },
+            "Value must be an integer or integer string.",
+            1,
+            id="disallow floats",
```
Comment (jusiskin) [inline test/openjd/model/v2023_09/test_chunk_int_task_parameter_type.py:309]: nit: should this mention "range"? same applies to test cases below this

File: src/openjd/model/v2023_09/_model.py [hunk 6]
```diff
@@ -797,11 +851,134 @@ class PathTaskParameterDefinition(OpenJDModel_v2023_09):
...
+                                    type="value_error",
+                                    loc=(i,),
+                                    ctx={
+                                        "error": ValueError(
+                                            "String literal must contain an integer."
+                                        )
+                                    },
+                                    input=item,
+                                )
+                            )
```
Comment (jusiskin) [inline src/openjd/model/v2023_09/_model.py:961]: nit: this code comment doesn't match the implementation. It seems we are only rejecting strings if they have no expressions and the string cannot be parsed as an integer.

**Review by jusiskin — COMMENTED** [review]
Great job. Just a few nitpicks, but I don't consider them blocking. Feel free to address them or merge as-is.

**Review by jusiskin — APPROVED** [review]

**Comment by mwiebe** [comment]
> Great job. Just a few nitpicks, but I don't consider them blocking. Feel free to address them or merge as-is.

Thanks! I'll merge this and then open a smaller PR to fix up what you found.
